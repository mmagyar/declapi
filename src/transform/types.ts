import Ajv from 'ajv'
import { ObjectType } from 'yaschva'
import { Elastic, KeyValue } from '../DataDriverTypes.js'
import { HttpMethods, SearchTypes } from '../globalTypes.js'
import { dirname } from 'path'
import { fileURLToPath } from 'url'

export type AuthType = (string | {createdBy: boolean})[] | boolean
export type ManageableFields ={ createdBy?: boolean }

export type Contract = {
  name: string;
  type?: HttpMethods;
  authentication: AuthType;
  manageFields: ManageableFields;
  arguments: ObjectType;
  returns: ObjectType;
  description?: string;
};
export type CrudAuthAll = {
  get: AuthType;
  put: AuthType;
  post: string[] | boolean;
  delete: AuthType;
};

export type CrudAuthSome = {
  get: AuthType;
  modify: AuthType;
  delete?: AuthType;
};

export type CrudContract = {
  name: string;
  methods?: { get?:boolean, post?:boolean, put?: boolean, patch?:boolean, delete?:boolean},
  authentication: AuthType | CrudAuthAll | CrudAuthSome;
  manageFields?:ManageableFields;
  dataType: ObjectType;
  search?: SearchTypes;
  description?: string;
  preferredImplementation?: {type: 'elasticsearch'; index: string};
};

export type OutputSuccess = {
  name: string;
  authentication: AuthType;
  manageFields: ManageableFields;
  method: HttpMethods;
  arguments: ObjectType;
  returns: ObjectType;
  preferredImplementation?: Elastic | KeyValue;
  search?: SearchTypes;
};

export type Output =
  | {type: 'result'; key: string; results: OutputSuccess[]; errors?: undefined;}
  | {type: 'error'; key?: undefined; errors: (Ajv.ErrorObject | string)[] | string; results?: undefined;};

export const baseSchemaLocation = `${dirname(fileURLToPath(import.meta.url))}/../../schema/`
