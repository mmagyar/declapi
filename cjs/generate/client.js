"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.client = void 0;
const common_js_1 = require("./common.js");
const getTokenDefaultPath = '../getToken';
const containsOptional = (input) => (Array.isArray(input) && input.some(y => y === '?')) ||
    input === '?';
const allOptional = (input) => Object.values(input).every(containsOptional);
exports.client = (input, getTokenPath) => {
    const functions = input.map(x => {
        const url = x.method === 'GET' ? `\`/api/${x.name}?\${serialize(input || {})}\`` : `"/api/${x.name}"`;
        const fn = `
    const valid = validate(contracts.${common_js_1.name(x)}.arguments, input);
      if (valid.result === "fail") throw valid;
    return fetch(${url}, {
      method: "${x.method}",
      headers: await defaultHeader(),
      body: ${x.method === 'GET' ? 'undefined' : 'JSON.stringify(input)'},
      signal
    }).then(x=> x.json())
    .then(x=> {
      if(x.errorType){throw x}
      const r = validate(contracts.${common_js_1.name(x)}.returns, x);
      if(r.result ==="fail") throw r; return x})\n`;
        return `export const ${common_js_1.name(x)} =
      async (input: ${common_js_1.name(x)}Argument${allOptional(x.arguments) ? ' = {}' : ''},
        signal: AbortSignal | undefined = undefined):
       Promise<${common_js_1.name(x)}Returns> => {\n${fn}};\n`;
    }).join('\n');
    const contractDescription = input.map(x => `${common_js_1.name(x)}: {
    name: "${x.name}",
    authentication: ${JSON.stringify(x.authentication, undefined, 2)},
    type: "${x.method}",
    arguments: ${JSON.stringify(x.arguments)},
    returns: ${JSON.stringify(x.returns)}}`).join(',');
    const fileContent = `/* eslint-disable max-lines */
/**********************************************
  DO NOT EDIT THIS FILE, IT WILL BE OVERRIDDEN
***********************************************/

  import { validate, Validation } from "yaschva";
  import { deepFreeze, serialize } from "microtil";
  import getToken from "${getTokenPath !== null && getTokenPath !== void 0 ? getTokenPath : getTokenDefaultPath}"
  const defaultHeader = async () => {
    const token = await getToken();
    const headers =  {
      'Authorization': \`Bearer \${token}\`,
      'Accept': 'application/json',
      'Content-Type': 'application/json'
    }

    if(!token) delete headers.Authorization
    return headers;
  };

    export const contracts: {[key:string]:{
      name:string, type:string, authentication:any,
      arguments: Validation, returns:Validation}} =
        deepFreeze({${contractDescription}});

    ${common_js_1.typeDef(input)}
    ${functions}
    export const functions = {${input.map(x => `${common_js_1.name(x)}: ${common_js_1.name(x)}`)}}
  `;
    return fileContent;
};
exports.default = exports.client;
//# sourceMappingURL=client.js.map