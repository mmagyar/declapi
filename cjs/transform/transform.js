"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.transform = void 0;
const crud_js_1 = require("./crud.js");
const single_js_1 = require("./single.js");
exports.transform = (contract) => __awaiter(void 0, void 0, void 0, function* () {
    const data = contract;
    if (!data.$schema) {
        return { type: 'error', errors: "Schema files must contain $schema that point to it's type" };
    }
    if (data.$schema.endsWith('singleContractSchema.json')) {
        return single_js_1.transform(data);
    }
    else if (data.$schema.endsWith('crudContractSchema.json')) {
        return crud_js_1.transform(data);
    }
    return { type: 'error', errors: `Unsupported schema for declaration: ${data.$schema}` };
});
exports.default = exports.transform;
//# sourceMappingURL=transform.js.map