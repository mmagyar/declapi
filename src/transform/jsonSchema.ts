import Ajv, { ErrorObject } from 'ajv'
export const jsonValidatorInit = () => {
  return new Ajv({
    loadSchema: async (uri: any) => require(uri.replace('node_modules/', ''))
  })
}

export type jsonValidationSuccess = {type:'success', errors? : ErrorObject[]};
export type jsonValidationError = {type:'error', errors : ErrorObject[]};
export type validateResult = jsonValidationSuccess | jsonValidationError;
export const isValidationError = (input:validateResult) : input is jsonValidationError => input && input.type === 'error'
export const validate = async (json: Object, data:any):
 Promise<validateResult> => {
  const validator = await jsonValidatorInit().compileAsync(json)
  const result = validator(data)

  const errors = validator.errors || []
  if (!result) return { type: 'error', errors }
  return { type: 'success' }
}
