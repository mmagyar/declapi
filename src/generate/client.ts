import { OutputSuccess } from '../transform/types'
import { typeDef, name } from './common'
import { Validation } from 'yaschva'

const getTokenDefaultPath = '../getToken'
const containsOptional = (input: Validation) =>
  (Array.isArray(input) && input.some(y => y === '?')) ||
   input === '?'

const allOptional = (input: Validation) =>
  Object.values(input).every(containsOptional)

export const client = (input: OutputSuccess[], getTokenPath?:string): string => {
  const functions = input.map(x => {
    const url =
      x.method === 'get' ? `\`/api/${x.name}?\${serialize(input || {})}\`` : `"/api/${x.name}"`
    const fn = `
    const valid = validate(contracts.${name(x)}.arguments, input);
      if (valid.result === "fail") throw valid;
    return fetch(${url}, {
      method: "${x.method}",
      headers: await defaultHeader(),
      body: ${x.method === 'get' ? 'undefined' : 'JSON.stringify(input)'},
      signal
    }).then(x=> x.json())
    .then(x=> {
      if(x.errorType){throw x}
      const r = validate(contracts.${name(x)}.returns, x);
      if(r.result ==="fail") throw r; return x})\n`

    return `export const ${name(x)} =
      async (input: ${name(x)}Argument${allOptional(x.arguments) ? ' = {}' : ''},
        signal: AbortSignal | undefined = undefined):
       Promise<${name(x)}Returns> => {\n${fn}};\n`
  }).join('\n')

  const contractDescription = input.map(x => `${name(x)}: {
    name: "${x.name}",
    authentication: ${JSON.stringify(x.authentication, undefined, 2)},
    type: "${x.method}",
    arguments: ${JSON.stringify(x.arguments)},
    returns: ${JSON.stringify(x.returns)}}`).join(',')

  const fileContent = `/* eslint-disable max-lines */
/**********************************************
  DO NOT EDIT THIS FILE, IT WILL BE OVERRIDDEN
***********************************************/

  import { validate, Validation } from "yaschva";
  import { deepFreeze, serialize } from "microtil";
  import getToken from "${getTokenPath ?? getTokenDefaultPath}"
  const defaultHeader = async () => {
    const token = await getToken();
    const headers =  {
      'Authorization': \`Bearer \${token}\`,
      'Accept': 'application/json',
      'Content-Type': 'application/json'
    }

    if(!token) delete headers.Authorization
    return headers;
  };

    export const contracts: {[key:string]:{
      name:string, type:string, authentication:any,
      arguments: Validation, returns:Validation}} =
        deepFreeze({${contractDescription}});

    ${typeDef(input)}
    ${functions}
    export const functions = {${input.map(x => `${name(x)}: ${name(x)}`)}}
  `
  return fileContent
}

export default client
