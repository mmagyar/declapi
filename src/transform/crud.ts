import { ValueType, ObjectType, StringType } from 'yaschva'
import { map } from 'microtil'
import { validate, isValidationError } from './jsonSchema'
import {
  CrudContract,

  CrudAuthAll,
  CrudAuthSome, OutputSuccess, Output, baseSchemaLocation, ManageableFields
} from './types'
import {
  HttpMethods,
  SearchTypes
} from 'declarapi-runtime'

const contractOptions = (input: ValueType | ValueType[]): ValueType[] => {
  if (Array.isArray(input)) {
    if (input.some(x => x === '?')) { return input }

    return input.concat(['?'])
  }
  return [input, '?']
}

const searchToType =
  (idType: 'string'| StringType, dataType: ObjectType, search?: SearchTypes): ObjectType => {
    if (search === 'idOnly') {
      return { id: [idType, { $array: idType }, '?'] }
    } else if (search === 'textSearch') {
      return { search: ['string', '?'], id: [idType, { $array: idType }, '?'] }
    } else if (search === 'full') {
      return map(dataType, value => contractOptions(value))
    } else if (!search) {
      return {}
    }

    return search
  }

const checkIdField = (contract:CrudContract) :Output | false => {
  if (contract.dataType.id === undefined) {
    return {
      type: 'error',
      errors: 'id field does not exist in the data declaration'
    }
  }

  const idType: any = contract.dataType.id
  if (!(idType === 'string' || idType.$string)) {
    return {
      type: 'error',
      errors: 'Type of id field must be string'
    }
  }

  return false
}
const checkManageFields = (contract: CrudContract): Output|false => {
  for (const [key, value] of Object.entries(contract.manageFields || {})) {
    if (value) {
      const fieldType: any = contract.dataType[key]
      if (fieldType === undefined) {
        return {
          type: 'error',
          errors: `managed field "${key}" is not present on data type`
        }
      }

      if (!(fieldType === 'string' || fieldType.$string)) {
        return {
          type: 'error',
          errors: `managed field "${key}" must be a string, current type :${fieldType}`
        }
      }
    }
  }

  return false
}

const removeManaged = (args: ObjectType, manageFields?: ManageableFields):ObjectType => {
  const result = { ...args }
  for (const [key, value] of Object.entries(manageFields || {})) {
    if (value) {
      delete result[key]
    }
  }

  return result
}
const isCrudAuth = (tbd: any): tbd is CrudAuthAll => tbd.post !== undefined
const isCrudAuthSome = (tbd: any): tbd is CrudAuthSome => tbd.modify !== undefined
const transformForPost = (tbd: any) => Array.isArray(tbd) && tbd.find(x => x.createdBy) ? true : tbd
export const transform = async (data:CrudContract | any): Promise<Output> => {
  const valid = await validate(require(`${baseSchemaLocation}crudContractSchema.json`), data)
  if (isValidationError(valid)) return valid

  const contractData: CrudContract = data

  const au = contractData.authentication

  const auth = {
    GET: isCrudAuth(au) ? au.get : (isCrudAuthSome(au) ? au.get : au),
    POST: isCrudAuth(au) ? au.post : transformForPost((isCrudAuthSome(au) ? au.modify : au)),
    PUT: isCrudAuth(au) ? au.put : (isCrudAuthSome(au) ? au.modify : au),
    PATCH: isCrudAuth(au) ? au.put : (isCrudAuthSome(au) ? au.modify : au),
    DELETE: isCrudAuth(au) ? au.delete : (isCrudAuthSome(au) ? au.delete || au.modify : au)
  }

  const createOutput = (method: HttpMethods, args: ObjectType,
    returns: ObjectType = contractData.dataType): OutputSuccess => ({
    method,
    name: contractData.name,
    authentication: auth[method],
    manageFields: contractData.manageFields || {},
    search: contractData.search,
    preferredImplementation: contractData.preferredImplementation,
    arguments: args,
    returns
  })

  const returnArray = { $array: contractData.dataType }
  const output: OutputSuccess[] = []

  const errorWithId = checkIdField(contractData)
  if (errorWithId) return errorWithId

  const errorWithManageFields = checkManageFields(contractData)
  if (errorWithManageFields) return errorWithManageFields

  const idType: any = contractData.dataType.id
  if (contractData.methods?.get !== false) {
    const search = contractData.search
    output.push(createOutput('GET',
      searchToType(idType, contractData.dataType, search), returnArray))
  }

  if (contractData.methods?.post !== false) {
    const post = { ...contractData.dataType, id: [idType, '?'] }
    output.push(createOutput('POST', removeManaged(post, contractData.manageFields)))
  }

  if (contractData.methods?.put !== false) {
    output.push(createOutput('PUT', removeManaged(contractData.dataType, contractData.manageFields)))
  }

  if (contractData.methods?.patch !== false) {
    const patch: {[s: string]: ValueType | ValueType[];} =
    { ...map(contractData.dataType, contractOptions), id: idType }
    output.push(createOutput('PATCH', removeManaged(patch, contractData.manageFields)))
  }

  if (contractData.methods?.delete !== false) {
    const deleteIds: {[s: string]: ValueType[];} = { id: [idType, { $array: idType }] }
    output.push(createOutput('DELETE', deleteIds, returnArray))
  }

  return { type: 'result', key: contractData.name, results: output }
}

export default transform
