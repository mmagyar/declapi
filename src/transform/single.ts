import { validate, isValidationError } from './jsonSchema.js'
import { Contract, Output, baseSchemaLocation } from './types.js'
import { loadJSON } from '../util.js'

export const transform = async (data:Contract | any): Promise<Output> => {
  const valid = await validate(await loadJSON(`${baseSchemaLocation}singleContractSchema.json`), data)
  if (isValidationError(valid)) return valid
  const contractData: Contract = data
  return {
    type: 'result',
    key: contractData.name,
    results: [
      {
        name: contractData.name,
        authentication: contractData.authentication,
        manageFields: {},
        method: contractData.type || 'get',
        arguments: data.arguments,
        returns: data.returns
      }
    ]
  }
}

export default transform
