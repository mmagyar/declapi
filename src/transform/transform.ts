import { Output, CrudContract, Contract } from './types'
import { transform as crudTransform } from './crud'
import { transform as singleTransform } from './single'

export const transform = async (contract: CrudContract | Contract|string | object): Promise<Output> => {
  const data: any = typeof contract === 'string' ? require(contract) : contract
  if (!data.$schema) { return { type: 'error', errors: "Schema files must contain $schema that point to it's type" } }

  if (data.$schema.endsWith('singleContractSchema.json')) {
    return singleTransform(data)
  } else if (data.$schema.endsWith('crudContractSchema.json')) {
    return crudTransform(data)
  }
  return { type: 'error', errors: `Unsupported schema for declaration: ${data.$schema}` }
}

export default transform