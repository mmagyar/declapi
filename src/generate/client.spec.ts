import client from './client'
import { OutputSuccess } from '../transform/types'
describe('Generate typing and fetch function for client', () => {
  const singleExample = (): OutputSuccess[] => [
    {
      name: 'test',
      authentication: false,
      method: 'get',
      arguments: { myNumber: 'number' },
      returns: {}
    }
  ]

  const crudExample = (): OutputSuccess[] => [
    {
      method: 'get',
      name: 'test',
      authentication: false,
      arguments: { search: ['string', '?'], id: ['string', { $array: 'string' }, '?'] },
      returns: { $array: { id: 'string', myNumber: 'number' } }
    },
    {
      method: 'post',
      name: 'test',
      authentication: false,
      arguments: { id: ['string', '?'], myNumber: 'number' },
      returns: { id: 'string', myNumber: 'number' }
    },
    {
      method: 'put',
      name: 'test',
      authentication: false,
      arguments: { id: 'string', myNumber: 'number' },
      returns: { id: 'string', myNumber: 'number' }
    },
    {
      method: 'patch',
      name: 'test',
      authentication: false,
      arguments: { id: 'string', myNumber: ['number', '?'] },
      returns: { id: 'string', myNumber: 'number' }
    },
    {
      method: 'delete',
      name: 'test',
      authentication: false,
      arguments: { id: ['string', { $array: 'string' }] },
      returns: { $array: { id: 'string', myNumber: 'number' } }
    }
  ]

  it('Generates single example without an error that matches snapshot', () => {
    const result = client(singleExample())
    expect(result).toMatchSnapshot()
  })

  it('Generates crud example without an error that matches snapshot', () => {
    const result = client(crudExample())
    expect(result).toMatchSnapshot()
  })

  it('Can set the import path for getToken function', () => {
    const result = client(singleExample(), '../myCustomTokenPath')
    expect(result).toMatchSnapshot()
  })
})
