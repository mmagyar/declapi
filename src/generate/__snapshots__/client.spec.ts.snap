// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Generate typing and fetch function for client Can set the import path for getToken function 1`] = `
"/* eslint-disable max-lines */
/**********************************************
  DO NOT EDIT THIS FILE, IT WILL BE OVERRIDDEN
***********************************************/

  import { validate, Validation } from \\"yaschva\\";
  import { deepFreeze, serialize } from \\"microtil\\";
  import getToken from \\"../myCustomTokenPath\\"
  const defaultHeader = async () => {
    const token = await getToken();
    const headers =  {
      'Authorization': \`Bearer \${token}\`,
      'Accept': 'application/json',
      'Content-Type': 'application/json'
    }

    if(!token) delete headers.Authorization
    return headers;
  };

    export const contracts: {[key:string]:{
      name:string, type:string, authentication:any,
      arguments: Validation, returns:Validation}} =
        deepFreeze({testGet: {
    name: \\"test\\",
    authentication: false,
    type: \\"get\\",
    arguments: {\\"myNumber\\":\\"number\\"},
    returns: {}}});

    /** test types for get method **/
export type testGetArgument = { myNumber: number }
export type testGetReturns = {  } | undefined

    export const testGet =
      async (input: testGetArgument,
        signal: AbortSignal | undefined = undefined):
       Promise<testGetReturns> => {

    const valid = validate(contracts.testGet.arguments, input);
      if (valid.result === \\"fail\\") throw valid;
    return fetch(\`/api/test?\${serialize(input || {})}\`, {
      method: \\"get\\",
      headers: await defaultHeader(),
      body: undefined,
      signal
    }).then(x=> x.json())
    .then(x=> {
      if(x.errorType){throw x}
      const r = validate(contracts.testGet.returns, x);
      if(r.result ===\\"fail\\") throw r; return x})
};

    export const functions = {testGet: testGet}
  "
`;

exports[`Generate typing and fetch function for client Generates crud example without an error that matches snapshot 1`] = `
"/* eslint-disable max-lines */
/**********************************************
  DO NOT EDIT THIS FILE, IT WILL BE OVERRIDDEN
***********************************************/

  import { validate, Validation } from \\"yaschva\\";
  import { deepFreeze, serialize } from \\"microtil\\";
  import getToken from \\"../getToken\\"
  const defaultHeader = async () => {
    const token = await getToken();
    const headers =  {
      'Authorization': \`Bearer \${token}\`,
      'Accept': 'application/json',
      'Content-Type': 'application/json'
    }

    if(!token) delete headers.Authorization
    return headers;
  };

    export const contracts: {[key:string]:{
      name:string, type:string, authentication:any,
      arguments: Validation, returns:Validation}} =
        deepFreeze({testGet: {
    name: \\"test\\",
    authentication: false,
    type: \\"get\\",
    arguments: {\\"search\\":[\\"string\\",\\"?\\"],\\"id\\":[\\"string\\",{\\"$array\\":\\"string\\"},\\"?\\"]},
    returns: {\\"$array\\":{\\"id\\":\\"string\\",\\"myNumber\\":\\"number\\"}}},testPost: {
    name: \\"test\\",
    authentication: false,
    type: \\"post\\",
    arguments: {\\"id\\":[\\"string\\",\\"?\\"],\\"myNumber\\":\\"number\\"},
    returns: {\\"id\\":\\"string\\",\\"myNumber\\":\\"number\\"}},testPut: {
    name: \\"test\\",
    authentication: false,
    type: \\"put\\",
    arguments: {\\"id\\":\\"string\\",\\"myNumber\\":\\"number\\"},
    returns: {\\"id\\":\\"string\\",\\"myNumber\\":\\"number\\"}},testPatch: {
    name: \\"test\\",
    authentication: false,
    type: \\"patch\\",
    arguments: {\\"id\\":\\"string\\",\\"myNumber\\":[\\"number\\",\\"?\\"]},
    returns: {\\"id\\":\\"string\\",\\"myNumber\\":\\"number\\"}},testDelete: {
    name: \\"test\\",
    authentication: false,
    type: \\"delete\\",
    arguments: {\\"id\\":[\\"string\\",{\\"$array\\":\\"string\\"}]},
    returns: {\\"$array\\":{\\"id\\":\\"string\\",\\"myNumber\\":\\"number\\"}}}});

    /** test types for get method **/
export type testGetArgument = { search?: string | undefined; id?: string | string[] | undefined } | undefined
export type testGetReturns = { id: string; myNumber: number }[]

/** test types for post method **/
export type testPostArgument = { id?: string | undefined; myNumber: number }
export type testPostReturns = { id: string; myNumber: number }

/** test types for put method **/
export type testPutArgument = { id: string; myNumber: number }
export type testPutReturns = { id: string; myNumber: number }

/** test types for patch method **/
export type testPatchArgument = { id: string; myNumber?: number | undefined }
export type testPatchReturns = { id: string; myNumber: number }

/** test types for delete method **/
export type testDeleteArgument = { id: string | string[] }
export type testDeleteReturns = { id: string; myNumber: number }[]

    export const testGet =
      async (input: testGetArgument = {},
        signal: AbortSignal | undefined = undefined):
       Promise<testGetReturns> => {

    const valid = validate(contracts.testGet.arguments, input);
      if (valid.result === \\"fail\\") throw valid;
    return fetch(\`/api/test?\${serialize(input || {})}\`, {
      method: \\"get\\",
      headers: await defaultHeader(),
      body: undefined,
      signal
    }).then(x=> x.json())
    .then(x=> {
      if(x.errorType){throw x}
      const r = validate(contracts.testGet.returns, x);
      if(r.result ===\\"fail\\") throw r; return x})
};

export const testPost =
      async (input: testPostArgument,
        signal: AbortSignal | undefined = undefined):
       Promise<testPostReturns> => {

    const valid = validate(contracts.testPost.arguments, input);
      if (valid.result === \\"fail\\") throw valid;
    return fetch(\\"/api/test\\", {
      method: \\"post\\",
      headers: await defaultHeader(),
      body: JSON.stringify(input),
      signal
    }).then(x=> x.json())
    .then(x=> {
      if(x.errorType){throw x}
      const r = validate(contracts.testPost.returns, x);
      if(r.result ===\\"fail\\") throw r; return x})
};

export const testPut =
      async (input: testPutArgument,
        signal: AbortSignal | undefined = undefined):
       Promise<testPutReturns> => {

    const valid = validate(contracts.testPut.arguments, input);
      if (valid.result === \\"fail\\") throw valid;
    return fetch(\\"/api/test\\", {
      method: \\"put\\",
      headers: await defaultHeader(),
      body: JSON.stringify(input),
      signal
    }).then(x=> x.json())
    .then(x=> {
      if(x.errorType){throw x}
      const r = validate(contracts.testPut.returns, x);
      if(r.result ===\\"fail\\") throw r; return x})
};

export const testPatch =
      async (input: testPatchArgument,
        signal: AbortSignal | undefined = undefined):
       Promise<testPatchReturns> => {

    const valid = validate(contracts.testPatch.arguments, input);
      if (valid.result === \\"fail\\") throw valid;
    return fetch(\\"/api/test\\", {
      method: \\"patch\\",
      headers: await defaultHeader(),
      body: JSON.stringify(input),
      signal
    }).then(x=> x.json())
    .then(x=> {
      if(x.errorType){throw x}
      const r = validate(contracts.testPatch.returns, x);
      if(r.result ===\\"fail\\") throw r; return x})
};

export const testDelete =
      async (input: testDeleteArgument,
        signal: AbortSignal | undefined = undefined):
       Promise<testDeleteReturns> => {

    const valid = validate(contracts.testDelete.arguments, input);
      if (valid.result === \\"fail\\") throw valid;
    return fetch(\\"/api/test\\", {
      method: \\"delete\\",
      headers: await defaultHeader(),
      body: JSON.stringify(input),
      signal
    }).then(x=> x.json())
    .then(x=> {
      if(x.errorType){throw x}
      const r = validate(contracts.testDelete.returns, x);
      if(r.result ===\\"fail\\") throw r; return x})
};

    export const functions = {testGet: testGet,testPost: testPost,testPut: testPut,testPatch: testPatch,testDelete: testDelete}
  "
`;

exports[`Generate typing and fetch function for client Generates single example without an error that matches snapshot 1`] = `
"/* eslint-disable max-lines */
/**********************************************
  DO NOT EDIT THIS FILE, IT WILL BE OVERRIDDEN
***********************************************/

  import { validate, Validation } from \\"yaschva\\";
  import { deepFreeze, serialize } from \\"microtil\\";
  import getToken from \\"../getToken\\"
  const defaultHeader = async () => {
    const token = await getToken();
    const headers =  {
      'Authorization': \`Bearer \${token}\`,
      'Accept': 'application/json',
      'Content-Type': 'application/json'
    }

    if(!token) delete headers.Authorization
    return headers;
  };

    export const contracts: {[key:string]:{
      name:string, type:string, authentication:any,
      arguments: Validation, returns:Validation}} =
        deepFreeze({testGet: {
    name: \\"test\\",
    authentication: false,
    type: \\"get\\",
    arguments: {\\"myNumber\\":\\"number\\"},
    returns: {}}});

    /** test types for get method **/
export type testGetArgument = { myNumber: number }
export type testGetReturns = {  } | undefined

    export const testGet =
      async (input: testGetArgument,
        signal: AbortSignal | undefined = undefined):
       Promise<testGetReturns> => {

    const valid = validate(contracts.testGet.arguments, input);
      if (valid.result === \\"fail\\") throw valid;
    return fetch(\`/api/test?\${serialize(input || {})}\`, {
      method: \\"get\\",
      headers: await defaultHeader(),
      body: undefined,
      signal
    }).then(x=> x.json())
    .then(x=> {
      if(x.errorType){throw x}
      const r = validate(contracts.testGet.returns, x);
      if(r.result ===\\"fail\\") throw r; return x})
};

    export const functions = {testGet: testGet}
  "
`;
